src/readCwd.cpp

```diff
@@ -0,0 +1,263 @@
+#include <string>
+
+#include <napi.h>
+
+#include <windows.h>
+#include <winternl.h>
+
+#define PTR_ADD_OFFSET(Pointer, Offset) ((PVOID)((ULONG_PTR)(Pointer) + (ULONG_PTR)(Offset)))
+
+struct CURDIR {
+  UNICODE_STRING DosPath;
+  HANDLE Handle;
+};
+
+struct UNICODE_STRING32 {
+  USHORT Length;
+  USHORT MaximumLength;
+  ULONG Buffer;
+};
+
+struct CURDIR32 {
+  UNICODE_STRING32 DosPath;
+  ULONG Handle;
+};
+
+// https://www.geoffchappell.com/studies/windows/km/ntoskrnl/inc/api/pebteb/rtl_user_process_parameters.htm
+struct LOCAL_RTL_USER_PROCESS_PARAMETERS {
+  ULONG _1[4];
+  HANDLE _2;
+  ULONG _3;
+  HANDLE _4[3];
+  CURDIR CurrentDirectory;
+};
+
+struct LOCAL_RTL_USER_PROCESS_PARAMETERS32 {
+  ULONG _1[4];
+  ULONG /* HANDLE32 */ _2;
+  ULONG _3;
+  ULONG /* HANDLE32 */ _4[3];
+  CURDIR32 CurrentDirectory;
+};
+
+struct PEB32 {
+  BOOLEAN _1[4];
+  ULONG _2[3];
+  ULONG ProcessParameters;
+};
+
+typedef NTSTATUS(__stdcall *nt_close)(HANDLE);
+typedef NTSTATUS(__stdcall *nt_openProcess)(PHANDLE, ACCESS_MASK, POBJECT_ATTRIBUTES, CLIENT_ID *);
+typedef NTSTATUS(__stdcall *nt_queryInformationProcess)(HANDLE, PROCESSINFOCLASS, PVOID, ULONG, ULONG_PTR);
+typedef NTSTATUS(__stdcall *nt_readVirtualMemory)(HANDLE, PVOID, PVOID, SIZE_T, PSIZE_T);
+
+struct NtCall {
+    HMODULE p_ntdll;
+    nt_close close;
+    nt_openProcess openProcess;
+    nt_queryInformationProcess queryInformationProcess;
+    nt_readVirtualMemory readVirtualMemory;
+};
+
+Napi::Value throwError(Napi::Env env, Napi::Error err) {
+  err.ThrowAsJavaScriptException();
+  return env.Null();
+}
+
+class GetCwdOfProcess : public Napi::Addon<GetCwdOfProcess> {
+  NtCall nt;
+
+  void initializeNtCall() {
+    this->nt.p_ntdll = GetModuleHandleW(L"ntdll.dll");
+    this->nt.close = nt_close(GetProcAddress(this->nt.p_ntdll, "NtClose"));
```

Comment:
This syntax is a little odd to me and hard to parse, I would expect something like this instead for these:

```suggestion
    this->nt.close = (nt_close) GetProcAddress(this->nt.p_ntdll, "NtClose");
```

Reply:
Also we should probably be checking the return value for `nullptr`.

Reply:
The code is more or less copied and pasted form the old version, but you're right, we should be doing some better error handling to prevent from seg-faulting the node/electron process.